docs/man/man3/BmsDeviceUpdater.3
.TH "BmsDeviceUpdater" 3 "Wed Jan 11 2017" "Version 0.0" "RaceUp_ino_core" \" -*- nroff -*-
.ad l
.nh
.SH NAME
BmsDeviceUpdater \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <BmsDeviceUpdater\&.h>\fP
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "template<class T > static void \fBsendSegmentStatus\fP (String statusType, byte cell, byte segment, T value)"
.br
.ti -1c
.RI "template<class T > static void \fBsendUpdate\fP (byte typeValue, T value)"
.br
.ti -1c
.RI "template<class T > static void \fBsendStatus\fP (T value)"
.br
.ti -1c
.RI "static void \fBsendSegmentVoltage\fP (byte cell, byte segment, double value)"
.br
.ti -1c
.RI "static void \fBsendPackVoltageOfBmsDevice\fP (byte n)"
.br
.ti -1c
.RI "static void \fBsendSegmentTemperature\fP (byte cell, byte segment, double value)"
.br
.ti -1c
.RI "static void \fBsendPackTemperatureOfBmsDevice\fP (byte n)"
.br
.ti -1c
.RI "static String \fBgetSerialUpdate\fP ()"
.br
.ti -1c
.RI "static String \fBcommandToString\fP (int command)"
.br
.in -1c
.SH "Detailed Description"
.PP 
Copyright 2016 - 2017 RaceUp Team Unipd
.PP
Licensed under the Apache License, Version 2\&.0 (the 'License'); you may not use this file except in compliance with the License\&. You may obtain a copy of the License at
.PP
http://www.apache.org/licenses/LICENSE-2.0
.PP
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied\&. See the License for the specific language governing permissions and limitations under the License\&. Sends updates on bms device to serial 
.SH "Member Function Documentation"
.PP 
.SS "String BmsDeviceUpdater::commandToString (int command)\fC [static]\fP"
Translate enum command to a string 
.PP
\fBParameters:\fP
.RS 4
\fIcommand\fP command number (as from the previously defined enum) 
.RE
.PP
\fBReturns:\fP
.RS 4
string representing name of command 
.RE
.PP

.SS "String BmsDeviceUpdater::getSerialUpdate ()\fC [static]\fP"
Reads serial data and checks if there is an update coming from the client 
.PP
\fBReturns:\fP
.RS 4
Raw String read from serial 
.RE
.PP

.SS "void BmsDeviceUpdater::sendPackTemperatureOfBmsDevice (byte n)\fC [static]\fP"
Send temperature data on the pack of bmsDevice device 
.PP
\fBParameters:\fP
.RS 4
\fIn\fP bmsDevice device 
.RE
.PP

.SS "void BmsDeviceUpdater::sendPackVoltageOfBmsDevice (byte n)\fC [static]\fP"
Send voltage data on the pack of bmsDevice device 
.PP
\fBParameters:\fP
.RS 4
\fIn\fP bmsDevice device 
.RE
.PP

.SS "template<class T > static void BmsDeviceUpdater::sendSegmentStatus (String statusType, byte cell, byte segment, T value)\fC [inline]\fP, \fC [static]\fP"
Send segment data 
.PP
\fBParameters:\fP
.RS 4
\fIstatusType\fP String about type of status to broadcast 
.br
\fIcell\fP number of cell 
.br
\fIsegment\fP number of segment 
.br
\fIvalue\fP value to send 
.RE
.PP

.SS "void BmsDeviceUpdater::sendSegmentTemperature (byte cell, byte segment, double value)\fC [static]\fP"
Send temperature data on the segment 
.PP
\fBParameters:\fP
.RS 4
\fIsender\fP Sender 
.br
\fIcell\fP Cell number 
.br
\fIsegment\fP Segment number 
.br
\fIsegmentVoltage\fP Temperature of segment in cell number 
.br
\fIvalue\fP value to send 
.RE
.PP

.SS "void BmsDeviceUpdater::sendSegmentVoltage (byte cell, byte segment, double value)\fC [static]\fP"
Send segment voltage data 
.PP
\fBParameters:\fP
.RS 4
\fIsender\fP Sender 
.br
\fIcell\fP Cell number 
.br
\fIsegment\fP Segment number 
.br
\fIsegmentVoltage\fP Voltage of segment in cell number 
.br
\fIvalue\fP value to send
.RE
.PP
Copyright 2016 - 2017 RaceUp Team Unipd
.PP
Licensed under the Apache License, Version 2\&.0 (the 'License'); you may not use this file except in compliance with the License\&. You may obtain a copy of the License at
.PP
http://www.apache.org/licenses/LICENSE-2.0
.PP
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied\&. See the License for the specific language governing permissions and limitations under the License\&. 
.SS "template<class T > static void BmsDeviceUpdater::sendStatus (T value)\fC [inline]\fP, \fC [static]\fP"
Send status 
.PP
\fBParameters:\fP
.RS 4
\fIsender\fP Sender 
.br
\fIvalue\fP Status to send 
.RE
.PP

.SS "template<class T > static void BmsDeviceUpdater::sendUpdate (byte typeValue, T value)\fC [inline]\fP, \fC [static]\fP"
Send data 
.PP
\fBParameters:\fP
.RS 4
\fIsender\fP Sender 
.br
\fItypeValue\fP Type of arg to send 
.br
\fIvalue\fP value to send 
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for RaceUp_ino_core from the source code\&.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## 0.2.1 - 2017-01-15
### Fixed
- non-OOP residue includes

## 0.2 - 2017-01-15
### Refactored
- OOP design implemented
### Removed
- useless classes like BmsDeviceReader, BmsDeviceWriter BmsDeviceUpdater

## 0.1.3 - 2017-01-11
### Added
- utils frame conversions

## 0.1.2 - 2017-01-11
### Added
- SdCardStream
### Tested
- compilation with Arduino UNO, DUE

## 0.1.1 - 2017-01-11
### Fixed
- ```#include``` bugs

## 0.1 - 2017-01-11
### Added
- Utils, BmsDevice, BmsDeviceReader, BmsDeviceUpdater, BmsDeviceWriter classes
- documentation
- data

## 0.0 - 2017-01-11
### Added
- project structure and repository


src/Bms/BmsDeviceUpdater.cpp
/**
 * Copyright 2016 - 2017 RaceUp Team Unipd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


#include "BmsDeviceUpdater.h"

void BmsDeviceUpdater::sendSegmentVoltage(byte cell, byte segment, double value) {
    sendSegmentStatus(commandToString(kCellVoltage), cell, segment, value);
}

void BmsDeviceUpdater::sendPackVoltageOfBmsDevice(byte n) {
    byte segment = (n - 1) / 3;
    for (byte cell = 0; cell < 18; cell++) {  // loop through cells monitored by bms device (6 of them)
        sendSegmentVoltage(cell, segment, BmsDeviceReader::getCellVoltage(n, cell));  // send voltage
    }
}

void BmsDeviceUpdater::sendSegmentTemperature(byte cell, byte segment, double value) {
    sendSegmentStatus(commandToString(kCellTemperature), cell, segment, value);
}

void BmsDeviceUpdater::sendPackTemperatureOfBmsDevice(byte n) {
    byte segment = (n - 1) / 3;
    for (byte cell = 0; cell < 18; cell++) {  // loop through cells monitored by bms device (6 of them)
        double value = BmsDeviceReader::getTemperature(n, TEMPERATURE1);
        sendSegmentTemperature(cell, segment, value);
    }
}

String BmsDeviceUpdater::getSerialUpdate() {
    return Serial.readString();
}

String BmsDeviceUpdater::commandToString(int command) {
    switch (command) {
        case 0:
            return (String) "Acknowledge";
        case 1:
            return (String) "Error";
        case 2:
            return (String) "Log";
        case 3:
            return (String) "Plot";
        case 4:
            return (String) "Alert";
        case 5:
            return (String) "Fault";
        case 6:
            return (String) "Status";
        case 7:
            return (String) "Voltage";
        case 8:
            return (String) "Temperature";
        case 9:
            return (String) "Sleep";
        default:
            return (String) "Unknown";
    }
}
